### PR TITLE
Implement long tweet serialization

### LINKED ISSUE #2: Split long messages into multiple tweets
### Description

If a messages is too long to tweet, currently it's just discarded. This should be split into multiple messages and tweeted as a thread. Hashtags will need to be considered as part of the length.

### PR DESCRIPTION
**Pull request by mpuckett159**
Implementation of serialization via tweet threads for longer messages being sent from Signal.

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -22,11 +22,11 @@ RUN wget -q https://github.com/AsamK/signal-cli/releases/download/v${SIGNAL_CLI_
     ln -sf /opt/signal-cli-${SIGNAL_CLI_VERSION}/bin/signal-cli /usr/local/bin/
 
 WORKDIR /app
-COPY scripts/register-number.sh /app
+COPY scripts/register-number.sh .
 COPY requirements.txt .
+COPY setup.cfg .
+COPY setup.py .
```
Comment (AetherUnbound) [inline Dockerfile:28]: Are these files necessary? Also I think it might be best (albeit redundant) to copy the `signal_scanner_bot` folder into the container in addition to having it mounted by the compose file. That way if we ever wanted to build/publish/ship the container directly it would have all of the code present on it. The mounting just allows us to restart the containers without having to rebuild the container.

File: docker-compose.yml [hunk 2]
```diff
@@ -8,7 +8,7 @@ services:
       - XDG_DATA_HOME=/app/data
     volumes:
       - signal-cli:/app/data/signal-cli
-      - .:/app
+      - ./signal_scanner_bot:/app/signal_scanner_bot
```
Comment (AetherUnbound) [inline docker-compose.yml:11]: Nice improvement! 🙂

File: signal_scanner_bot/twitter.py [hunk 3]
```diff
@@ -35,16 +36,73 @@ def get_api():
...
-    {tweet} @ {timestamp.strftime('%l:%M:%S%p').strip()}
-
-    {hashtags}
-    """
-    )
-    api.update_status(formatted)
+
+    # Check if tweet is longer than 280 minus a 20 char padding, add to list if
+    # not, continue with dividing up text if so
+    if len(tweet + hashtags) >= TWEET_MAX_SIZE - TWEET_PADDING:
```
Comment (AetherUnbound) [inline signal_scanner_bot/twitter.py:42]: I think it might be a good idea to move this logic out into a separate function that's present in this module. We could have a `send_single_tweet` function that still only ever sends one tweet, and then a `send_multiple_tweets` function that splits up the tweets and sends them out (or just receives a list of split up tweets). Then we can have a higher level `send_tweet` that accepts the signal message and dispatches to the appropriate method depending on message length. Thoughts?

File: signal_scanner_bot/twitter.py [hunk 4]
```diff
@@ -35,16 +36,73 @@ def get_api():
...
-    formatted = dedent(
-        f"""
-    {tweet} @ {timestamp.strftime('%l:%M:%S%p').strip()}
-
-    {hashtags}
-    """
-    )
-    api.update_status(formatted)
+
+    # Check if tweet is longer than 280 minus a 20 char padding, add to list if
```
Comment (AetherUnbound) [inline signal_scanner_bot/twitter.py:40]: LOVE LOVE LOVE these comments 💯

File: signal_scanner_bot/twitter.py [hunk 5]
```diff
@@ -35,16 +36,73 @@ def get_api():
...
+            elif index < len(tweet_word_list):
+                sub_tweet = " ".join(tweet_word_list[j:index]) + " ..."
+            elif index == len(tweet_word_list):
+                sub_tweet = " ".join(tweet_word_list[j:index])
+
+            # When length of tweet reaches >260 chars save to list and set
+            # base index for next tweet
+            if len(sub_tweet) > TWEET_MAX_SIZE - TWEET_PADDING:
+                tweets_list.append(" ".join(tweet_word_list[j:index - 1]) + " ...")
+                j = index - 1
```
Comment (AetherUnbound) [inline signal_scanner_bot/twitter.py:67]: Just so I understand, it looks like this section is iteratively building up the tweet (with any extra bits that are necessary depending on where it is in the thread) and then when the tweet exceeds the maximum, uses the last known working position to determine the tweet size?

File: signal_scanner_bot/twitter.py [hunk 6]
```diff
@@ -35,16 +36,73 @@ def get_api():
...
-    api.update_status(formatted)
+
+    # Check if tweet is longer than 280 minus a 20 char padding, add to list if
+    # not, continue with dividing up text if so
+    if len(tweet + hashtags) >= TWEET_MAX_SIZE - TWEET_PADDING:
+
+        # Split tweet into word list, create empty list to store serialized tweets, index to track last position
+        tweet_word_list = tweet.split(" ")
+        tweets_list = []
+        j = 0
```
Comment (AetherUnbound) [inline signal_scanner_bot/twitter.py:47]: I'm not the biggest fan of single letter variable names, could this be `start_index`?

File: signal_scanner_bot/twitter.py [hunk 7]
```diff
@@ -35,16 +36,73 @@ def get_api():
...
-    """
-    )
-    api.update_status(formatted)
+
+    # Check if tweet is longer than 280 minus a 20 char padding, add to list if
+    # not, continue with dividing up text if so
+    if len(tweet + hashtags) >= TWEET_MAX_SIZE - TWEET_PADDING:
+
+        # Split tweet into word list, create empty list to store serialized tweets, index to track last position
+        tweet_word_list = tweet.split(" ")
```
Comment (AetherUnbound) [inline signal_scanner_bot/twitter.py:45]: I didn't think about this when you were initially testing, but we should see how this processing handles newlines

File: signal_scanner_bot/twitter.py [hunk 8]
```diff
@@ -35,16 +36,73 @@ def get_api():
...
+        tweets_list = []
+        j = 0
+
+        # For loop through word list and enumerate the index because we'll need
+        # it but don't actually care about the list value so send to null
+        for index, _ in enumerate(tweet_word_list):
```
Comment (AetherUnbound) [inline signal_scanner_bot/twitter.py:51]: For this (and below on line 74) you can actually use the following logic without having to extract values from the list:
```suggestion
        for index in range(len(tweet_word_list)):
```

File: signal_scanner_bot/twitter.py [hunk 9]
```diff
@@ -35,16 +36,73 @@ def get_api():
...
+
+            {sub_tweet}
+
+            {hashtags}
+            """
+            )
+        else:
+            formatted = dedent(
+                f"{sub_tweet}"
+            )
```
Comment (AetherUnbound) [inline signal_scanner_bot/twitter.py:99]: I think this can just be the following:
```suggestion
            formatted = sub_tweet
```

File: signal_scanner_bot/twitter.py [hunk 10]
```diff
@@ -35,16 +36,73 @@ def get_api():
...
+
+        # If first tweet send without reply to tweet ID parameter, if part
+        # of a thread send reply using ID of last tweet sent.
+        if tweet_id is None:
+            status_obj = api.update_status(status=formatted)
+            tweet_id = status_obj._json['id']
```
Comment (AetherUnbound) [inline signal_scanner_bot/twitter.py:105]: You _should_ be able to do the following:
```suggestion
            status = api.update_status(status=formatted)
            tweet_id = status.id
```

File: signal_scanner_bot/twitter.py [hunk 11]
```diff
@@ -35,16 +36,73 @@ def get_api():
...
+        else:
+            status_obj = api.update_status(status=formatted, in_reply_to_status_id=tweet_id, auto_populate_reply_metadata=True)
+            tweet_id = status_obj._json['id']
```
Comment (AetherUnbound) [inline signal_scanner_bot/twitter.py:108]: Same here:
```suggestion
            status = api.update_status(status=formatted, in_reply_to_status_id=tweet_id, auto_populate_reply_metadata=True)
            tweet_id = status.id
```

File: signal_scanner_bot/twitter.py [hunk 12]
```diff
@@ -35,16 +36,73 @@ def get_api():
...
+            if len(sub_tweet) > TWEET_MAX_SIZE - TWEET_PADDING:
+                tweets_list.append(" ".join(tweet_word_list[j:index - 1]) + " ...")
+                j = index - 1
+
+        # Save the last tweet
+        tweets_list.append(" ".join(tweet_word_list[j:]))
+
+        # Append the tweet number / tweet thread length to end of tweet. Again
+        # don't actually care about the list value so sending to null.
+        for index, _ in enumerate(tweets_list):
```
Comment (AetherUnbound) [inline signal_scanner_bot/twitter.py:74]:
```suggestion
        for index in range(len(tweets_list)):
```

**Review by AetherUnbound — COMMENTED** [review]
Thanks a ton for this great contribution! I have a couple suggestions on style and structuring, but overall I think the flow/logic look sound. Also it looks like CI/CD is failing on a few things, let me know if you have any issues running those tools (you should be able to see them in the `.github/workflows/pull_request.yml` file if you want to know how to run them).

File: signal_scanner_bot/twitter.py (hunk 5 above)
Comment (mpuckett159) [reply signal_scanner_bot/twitter.py]: Yeah, the logic is a little wonky here honestly. I'm open to criticism and suggestions for improvement. 

I'm declaring the tweet_list as an empty array (will be updated to be an empty array of string for the final merge) and then storing then checking if that array is empty or not. If it is empty, that's the first tweet in the thread and will thus need the the hashtags in addition to the ellipses, if not and the index is less than the length of the total length of the array that will be an intermediary tweet and only need the ellipses, and finally when the index reaches the last point (ie. index equals length) then we are at the last tweet and thus don't need the ellipses anymore.

I realized this morning that I accidentally made it work out by not actually saving the last tweet, because the last tweet skips the:
if len(sub_tweet) > TWEET_MAX_SIZE - TWEET_PADDING:

Part, because in my testing I've just been using the same string, I think I'm missing a potential edge case where the last tweet I'm building would go over the max tweet length we allow on the last index, and then get into a weird state. I need to think about this and figure out how to improve this logic.

File: signal_scanner_bot/twitter.py (hunk 6 above)
Comment (mpuckett159) [reply signal_scanner_bot/twitter.py:47]: Updated to `base_index`

**Review by AetherUnbound — APPROVED** [review]
Looks great! Thanks again for this awesome change!
